### PR TITLE
refactor: unify upstream label to use config name

### PR DESCRIPTION
**Pull request by yaoge123**
## Background

When PR #34 was merged it added queue/upstream-level Prometheus metrics
that label connections by **upstream config name** (the `Name` of the
`[upstreams.<NAME>]` table), while the older per-connection metrics
added in PR #30 already labeled connections by the resolved **upstream
address** (`<host>:<port>` from `target.Addr`, plus `"unknown"`).

Both sets of series share the same label key `upstream` but their values
have completely different semantics, so dashboards cannot
`on(upstream)` join the two groups, and the same `upstream` value can
mean different things in different panels.

PR #30 only landed in the v0.1.4 → v0.1.5 cycle (no broad public
adoption yet), so cleaning this up now still has a low blast radius.

## Change

Switch the per-connection metrics to use the upstream's **config name**
as well, matching the queue/upstream metrics introduced in PR #34.

Affected metrics (label `upstream` value changes from
`<addr>` / `"unknown"` to the upstream's config Name):

- `rsync_proxy_active_connections_by_module`
- `rsync_proxy_connection_sent_bytes`
- `rsync_proxy_connection_received_bytes`
- `rsync_proxy_connection_connected_timestamp_seconds`
- `rsync_proxy_connection_duration_seconds`

The JSON tag on `ConnInfo.Upstream` is still `"upstream"` so the
`/connections` debug endpoint output keeps the same key for clients.

## Why config name

1. The config name is a stable logical identifier; the resolved address
   is an implementation detail that can shift on each pick.
2. When an upstream has multiple backends, aggregating by Name answers
   the natural question "how much traffic did this upstream handle?".
3. PR #34 already chose the Name route. Unifying lets dashboards join
   per-connection and queue metrics on `upstream`.

## Compatibility

This is a breaking change for anyone querying the per-connection
metrics by upstream **address** in dashboards/alerts. Series with
addresses will stop receiving data and new series with the upstream's
config Name will appear instead.

Given PR #30 only shipped in v0.1.5 and the metrics are intended for
operator dashboards (not external integrations), changing them now is
safer than leaving the inconsistency in place.

## Tests

- `go vet ./... && go build ./... && go test -race ./pkg/...` — all green.
- `TestStatusIncludesSelectedUpstream` and
  `TestMetricsIncludesActiveConnections` updated to assert the config
  Name; the latter also gained a `assert.NotContains` for the old address.

### CODE REVIEW

File: pkg/server/metrics.go [hunk 1]
```diff
@@ -110,7 +110,7 @@ func (s *Server) writePrometheusMetrics(w io.Writer, now time.Time) {
 		snapshot := conn.snapshot()
 		key := prometheusConnectionGroup{
 			module:   prometheusLabelValueOrUnknown(snapshot.Module),
-			upstream: prometheusLabelValueOrUnknown(snapshot.UpstreamAddr),
+			upstream: prometheusLabelValueOrUnknown(snapshot.Upstream),
 		}
```
[the remark this answers is not in the document]
Comment (yaoge123) [reply pkg/server/metrics.go:114]: Documented in the PR description: this is intentionally a breaking change to the per-connection metrics (`active_connections_by_module`, `connection_*`). The rationale is that the conflicting `upstream` semantics introduced in #30 vs #34 cannot be joined in PromQL, and #30 only landed in v0.1.5 with no broad public uptake yet, so cleaning it up now has a low blast radius. An additive migration with a parallel `upstream_id` label would freeze the inconsistency and double the metric footprint in every existing dashboard query — keeping the simpler unified label seems the better trade-off here. Happy to revisit if maintainers prefer the additive approach.

File: pkg/server/server.go [hunk 2]
```diff
@@ -59,7 +59,7 @@ type ConnInfo struct {
 	RemoteAddr    string
 	ConnectedAt   time.Time
 	Module        string
-	UpstreamAddr  string
+	Upstream      string
 	SentBytes     atomic.Int64
```
[the remark this answers is not in the document]
Comment (yaoge123) [reply pkg/server/server.go:63]: The field name `Upstream` is deliberately aligned with the existing `Target.Upstream` field on the same struct, the `[upstreams.<NAME>]` config table, and the `upstream` Prometheus label / JSON tag. Renaming the Go field to `UpstreamID` while keeping the JSON tag and metric label as `upstream` would create a Go-vs-wire mismatch, which seems worse than the current naming. The struct already exposes the resolved address separately as `Target.Addr` for callers that need it.

File: pkg/server/server_test.go [hunk 3]
```diff
@@ -375,7 +375,7 @@ func TestStatusIncludesSelectedUpstream(t *testing.T) {
 		if len(infos) != 1 {
 			return false
 		}
-		return infos[0].snapshot().UpstreamAddr == upstreamAddr
+		return infos[0].snapshot().Upstream == "u1"
```
[the remark this answers is not in the document]
Comment (yaoge123) [reply pkg/server/server_test.go:378]: The literal `"u1"` is the established convention across the existing test suite (see e.g. `TestPerUpstreamQueueIsolation`, `TestQueueFullRejectsConnection`, `TestMetricsCountsQueueFullRejection`); the new assertions just match it. Centralizing it in a single constant could make sense but is a separate, repo-wide refactor that would change a lot of unrelated tests, so leaving it consistent with the surrounding tests in this PR.

File: pkg/server/server_test.go [hunk 4]
```diff
@@ -445,7 +445,7 @@ func TestMetricsIncludesActiveConnections(t *testing.T) {
...
```
[the remark this answers is not in the document]
Comment (yaoge123) [reply pkg/server/server_test.go:448]: Same as the comment on line 378 — sticking with the existing `"u1"` convention used across the existing test file.

File: pkg/server/server_test.go [hunk 5]
```diff
@@ -458,14 +458,15 @@ func TestMetricsIncludesActiveConnections(t *testing.T) {
 
 	assert.Equal(t, http.StatusOK, resp.StatusCode)
 	assert.Contains(t, text, "rsync_proxy_active_connections 1\n")
-	assert.Contains(t, text, fmt.Sprintf("rsync_proxy_active_connections_by_module{module=\"fake\",upstream=%q} 1\n", upstreamAddr))
+	assert.Contains(t, text, "rsync_proxy_active_connections_by_module{module=\"fake\",upstream=\"u1\"} 1\n")
```
[the remark this answers is not in the document]
Comment (yaoge123) [reply pkg/server/server_test.go:461]: Same as the comment on line 378 — sticking with the existing `"u1"` convention used across the existing test file.

File: cmd/cmd.go [hunk 6]
```diff
@@ -137,7 +137,7 @@ func SendConnectionsRequest(addr string, stdout, stderr io.Writer) error {
 			strconv.Itoa(conn.Index),
 			conn.RemoteAddr,
 			conn.Module,
-			conn.UpstreamAddr,
+			conn.Upstream,
```
Comment (taoky) [inline cmd/cmd.go:140]: 这个看起来会改变 rsync-proxy connections 输出的表格内容，虽然我觉得问题不大

**Review by taoky — APPROVED** [review]

File: cmd/cmd.go (hunk 6 above)
Comment (iBug) [reply cmd/cmd.go:140]: 这个其实挺好的，我一直都想这么做

**Review by iBug — APPROVED** [review]
